src/primitives/box.jl

```diff
@@ -33,6 +33,23 @@ measure(b::Box) = prod(b.max - b.min)
 diagonal(b::Box) = norm(b.max - b.min)
 sides(b::Box) = b.max - b.min
 
+"""
+Return all corners of the box.
+"""
+@generated function vertices(b::Box{Dim,T}) where {Dim,T}
+  list = CartesianIndices(ntuple(i->2, Dim))
+  P = Point{Dim,T}
+  ex = Expr(:ref, P)
+  map(list) do idxs
+    next_tuple = Expr(:tuple, (:(coords[$(idxs[i])][$i]) for i in 1:Dim)...)
+    push!(ex.args, next_tuple)
+  end
+  quote
+    coords = coordinates(b.min), coordinates(b.max)
+    $ex
+  end
+end
```

Comment:
Again, I don't feel that this is the kind of code that is easy to read and maintain. Metaprogramming has its strengths but this is too big of a hammer to solve a very simple and specific feature with 2D and 3D boxes. You are trying to write code for a future we don't even know we will be able to support (higher dimensions). Let's keep it simple, please.

Suppose we share this code in any forum. Do you think people will know what it is doing in less than 1min? I am positive that people will struggle. Please rewrite it like we do in the CartesianGrid for example. Premature generalizations can be harmful too.

Reply:
Remember the [Zen of Python](https://www.python.org/dev/peps/pep-0020/#id2):

- Simple is better than complex.
- Explicit is better than implicit.
- Readability counts.
- Special cases aren't special enough to break the rules.
- Although practicality beats purity.

These are the main ones related to this issue.

Reply:
To be honest, the following snippet (from CartesianGrid) is not easy to understand for me. Locally it is, but I quickly lose sight of the global picture.

Snippet:
```julia
  if Dim == 1 # segment
    p1 = (o[1] + (i[1] - 1) * s[1],)
    p2 = (o[1] + (    i[1]) * s[1],)
    Segment(p1, p2)
  elseif Dim == 2 # quadrangle
    p1 = (o[1] + (i[1] - 1) * s[1],
          o[2] + (i[2] - 1) * s[2])
    p2 = (o[1] + (i[1]    ) * s[1],
          o[2] + (i[2] - 1) * s[2])
    p3 = (o[1] + (i[1]    ) * s[1],
          o[2] + (i[2]    ) * s[2])
    p4 = (o[1] + (i[1] - 1) * s[1],
          o[2] + (i[2]    ) * s[2])
    Quadrangle(p1, p2, p3, p4)
  elseif Dim == 3 # hexahedron
    p1 = (o[1] + (i[1] - 1) * s[1],
          o[2] + (i[2] - 1) * s[2],
          o[3] + (i[3] - 1) * s[3])
    p2 = (o[1] + (i[1]    ) * s[1],
          o[2] + (i[2] - 1) * s[2],
          o[3] + (i[3] - 1) * s[3])
    p3 = (o[1] + (i[1]    ) * s[1],
          o[2] + (i[2]    ) * s[2],
          o[3] + (i[3] - 1) * s[3])
    p4 = (o[1] + (i[1] - 1) * s[1],
          o[2] + (i[2]    ) * s[2],
          o[3] + (i[3] - 1) * s[3])
    p5 = (o[1] + (i[1] - 1) * s[1],
          o[2] + (i[2] - 1) * s[2],
          o[3] + (i[3]    ) * s[3])
    p6 = (o[1] + (i[1]    ) * s[1],
          o[2] + (i[2] - 1) * s[2],
          o[3] + (i[3]    ) * s[3])
    p7 = (o[1] + (i[1]    ) * s[1],
          o[2] + (i[2]    ) * s[2],
          o[3] + (i[3]    ) * s[3])
    p8 = (o[1] + (i[1] - 1) * s[1],
          o[2] + (i[2]    ) * s[2],
          o[3] + (i[3]    ) * s[3])
    Hexahedron(p1, p2, p3, p4, p5, p6, p7, p8)
  else
    throw(ErrorException("not implemented"))
  end
```

I brought a few changes to the generated function. It should be more readable now. I also added the manual versions. They are faster to process (like, you directly see the result), but it's more error-prone because you put yourself at the place of the machine and have to run the program in your head to confirm its validity instead of relying on structured logic. I still think that the generated function is better for a maintainability perspective, but if you still insist, I can delete it.